File: examples/word-add-comment.ts
/**
 * 例子（Epic-18）：从零构造一份 docx，在段落中间挂一条注释（commentRangeStart /
 * commentRangeEnd 包围目标文字，commentReference Run 放结尾）。
 *
 * 跑法：
 *   bun run examples/word-add-comment.ts <output.docx>
 */

import { Paragraph, Run, Text, WordprocessingDocument } from "../src/word/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: word-add-comment <output.docx>\n");
    process.exit(2);
  }

  const doc = WordprocessingDocument.create();
  const body = doc.mainDocumentPart!.document.firstChild()!;
  const append = (e: unknown): void =>
    (body as { appendChild: (e: unknown) => void }).appendChild(e);

  // 加注释 → 拿三个 markup 元素
  const { rangeStart, rangeEnd, reference, commentId } = doc.addComment({
    author: "Alice",
    initials: "AW",
    text: "This is an example comment.",
  });

  // 主文档段：前缀 Run + rangeStart + 目标 Run + rangeEnd + reference Run + 后缀 Run
  const p = new Paragraph();

  const lead = new Run();
  const leadT = new Text();
  leadT.text = "Hello ";
  lead.appendChild(leadT);
  p.appendChild(lead);

  p.appendChild(rangeStart);

  const target = new Run();
  const targetT = new Text();
  targetT.text = "world";
  target.appendChild(targetT);
  p.appendChild(target);

  p.appendChild(rangeEnd);
  p.appendChild(reference);

  const tail = new Run();
  const tailT = new Text();
  tailT.text = " — please review.";
  tail.appendChild(tailT);
  p.appendChild(tail);

  append(p);

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (comment id=${commentId} attached to "world")\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordAddComment.cs
// Replica of examples/word-add-comment.ts
// Creates a docx with a paragraph: "Hello [world] — please review."
// where "world" is annotated with a comment by Alice.

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;

namespace CrossSdkVerify.Replicas;

public static class WordAddComment
{
    public static void Run(string outputPath)
    {
        using var doc = WordprocessingDocument.Create(outputPath,
            DocumentFormat.OpenXml.WordprocessingDocumentType.Document);

        var mainPart = doc.AddMainDocumentPart();
        mainPart.Document = new Document();
        var body = new Body();
        mainPart.Document.Append(body);

        // Create word comments part
        var commentsPart = mainPart.AddNewPart<WordprocessingCommentsPart>();
        commentsPart.Comments = new Comments();

        // Comment content
        var comment = new Comment(
            new Paragraph(new Run(new Text("This is an example comment."))))
        {
            Id = "1",
            Author = "Alice",
            Initials = "AW",
            Date = new DateTime(2024, 1, 1, 0, 0, 0, DateTimeKind.Utc)
        };
        commentsPart.Comments.Append(comment);
        commentsPart.Comments.Save();

        // Main paragraph: "Hello " + rangeStart + "world" + rangeEnd + reference + " — please review."
        var p = new Paragraph();

        p.Append(new Run(new Text("Hello ") { Space = SpaceProcessingModeValues.Preserve }));
        p.Append(new CommentRangeStart { Id = "1" });
        p.Append(new Run(new Text("world")));
        p.Append(new CommentRangeEnd { Id = "1" });
        p.Append(new Run(new CommentReference { Id = "1" }));
        p.Append(new Run(new Text(" — please review.") { Space = SpaceProcessingModeValues.Preserve }));

        body.Append(p);
        body.Append(new SectionProperties());
        mainPart.Document.Save();
    }
}
